packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,162 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0 */
+import { FeatureFlags } from "aws-cdk-lib";
+import { Grant, IRole, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
+import { IBucket } from "aws-cdk-lib/aws-s3";
+import * as perms from "aws-cdk-lib/aws-s3/lib/perms";
+import * as cxapi from "aws-cdk-lib/cx-api";
+import { IConstruct } from "constructs";
+import { ErrorIntegrationResponse } from "./error-integration-response";
+import { ErrorIntegrationResponses } from "./error-integration-responses";
+import {
+  ApiGatewayIntegration,
+  Integration,
+  IntegrationGrantProps,
+  IntegrationRenderProps,
+} from "./integration";
+import { generateCorsResponseParameters } from "../prepare-spec-event-handler/prepare-spec";
+import { Method } from "../spec";
+import { bucketInvocationUri } from "../spec/utils";
+
+export interface S3IntegrationProps {
+  /**
+   * The S3 bucket to be invoked on integration
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * The HTTP method to use when invoking the S3 bucket
+   * @default - integration method is used
+   */
+  readonly method?: Method;
+
+  /**
+   * The path override to use when invoking the S3 bucket
+   * @default - integration path is used
+   */
+  readonly path?: string;
+
+  /**
+   * The error integration response to use when the S3 bucket returns an error
+   * @default ErrorIntegrationResponses.catchAll()
+   */
+  readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+}
+
+/**
+ * An S3 integration
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly method?: Method;
+  private readonly path?: string;
+  private readonly errorIntegrationResponse?: ErrorIntegrationResponse;
+
+  private readonly executionRoleId = "S3IntegrationsExecutionRole";
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+    this.method = props.method;
+    this.path = props.path;
+    this.errorIntegrationResponse = props.errorIntegrationResponse;
+  }
+
+  private executionRole(scope: IConstruct): IRole {
+    // Retrieve or create the shared S3 execution role
+    return (scope.node.tryFindChild(this.executionRoleId) ??
+      new Role(this.bucket, this.executionRoleId, {
```

Comment:
Ooh we should create the role on the passed in `scope` instead of the bucket :)

Reply:
🤦

Reply:
dumb mode activated